### ISSUE TITLE
How can I use custom head meta viewport  like user-scalable=no 

### ISSUE THREAD

**Issue by hbsion** [denoland/fresh#779]

**Comment by cnnrsprs** [comment]
I just picked up Fresh for the first time, came looking through the issues for the same reason and you happened to post one hour before me. Here's what I found (as a newbie):

Right now, it looks like Fresh exports the [`<Head>` component from the runtime module](https://deno.land/x/fresh@1.1.1/runtime.ts?doc=&s=Head) to solve this problem. This seems to work:

```ts
import { Head } from "$fresh/runtime.ts";

export function Home() {
  return (
    <div>
      <Head>
        <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no" />
      </Head>
      <main>content</main>
    </div>
  );
}
```

However, based on [the roadmap](https://github.com/denoland/fresh/issues/563), it appears the plan is to deprecate that in the near future and allow `<head>` anywhere without the need for an import.

Hope this helps.

**Comment by hbsion** [comment]
> I just picked up Fresh for the first time, came looking through the issues for the same reason and you happened to post one hour before me. Here's what I found (as a newbie):
> 
> Right now, it looks like Fresh exports the [`<Head>` component from the runtime module](https://deno.land/x/fresh@1.1.1/runtime.ts?doc=&s=Head) to solve this problem. This seems to work:
> 
> ```ts
> import { Head } from "$fresh/runtime.ts";
> 
> export function Home() {
>   return (
>     <div>
>       <Head>
>         <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no" />
>       </Head>
>       <main>content</main>
>     </div>
>   );
> }
> ```
> 
> However, based on [the roadmap](https://github.com/denoland/fresh/issues/563), it appears the plan is to deprecate that in the near future and allow `<head>` anywhere without the need for an import.
> 
> Hope this helps.

it is not working you can look this my code

> ```ts
> export default function Home() {
>   return (<>
>    <Head>
>      <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
>      <meta name="renderer" content="webkit"/>
  >    <meta name="force-rendering" content="webkit"/>
>      <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
>      <meta name="viewport" content="width=device-width,minimum-scale=1.0,maximum-scale=1.0,user-scalable=no" />
 >     <title>fresh demo</title>
 >   </Head>
 >   <div class="p-4 mx-auto max-w-screen-md h-screen box-border flex flex-col items-center justify-center gap-5">
  >    <img
>        src="/logo.svg"
>        class="w-32 h-32"
>        alt="the fresh logo: a sliced lemon dripping with juice"
>      />
>      <div class="text-gray-600">this is a demo website use deno fresh build</div>
>      <Counter start={3} />
 >     <div class="h-5"/>
>      <a href="https://fresh.deno.dev" target="_blank">
>        <img width="197" height="37" src="/fresh-badge.svg" alt="Made with Fresh" />
>      </a>
>      <div class="text-sm text-gray-400">&copy; {new Date().getFullYear()} xuwt.com</div>
>    </div>
>  </>);
>}
> ```

render html  is has <meta name="viewport" content="width=device-width, initial-scale=1.0" /> and <meta name="viewport" content="width=device-width,minimum-scale=1.0,maximum-scale=1.0,user-scalable=no" /> two viewport

**Comment by connorlogin** [comment]
I don't think anybody can help if you don't provide the code that generated the page in question, which would be much more helpful than posting an ambiguous off-site demo hosted by Huawei. Also, it looks like you're collecting IP addresses, which seems like a fishy thing to do while troubleshooting a github issue. I'm new here, but I imagine this isn't the best way to seek guidance.

The `<Head>` component works as expected in my dev setup.

**Comment by hbsion** [comment]
so sorry
this is my code

> ```ts
> export default function Home() {
>   return (<>
>    <Head>
>      <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
>      <meta name="renderer" content="webkit"/>
  >    <meta name="force-rendering" content="webkit"/>
>      <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
>      <meta name="viewport" content="width=device-width,minimum-scale=1.0,maximum-scale=1.0,user-scalable=no" />
 >     <title>fresh demo</title>
 >   </Head>
 >   <div class="p-4 mx-auto max-w-screen-md h-screen box-border flex flex-col items-center justify-center gap-5">
  >    <img
>        src="/logo.svg"
>        class="w-32 h-32"
>        alt="the fresh logo: a sliced lemon dripping with juice"
>      />
>      <div class="text-gray-600">this is a demo website use deno fresh build</div>
>      <Counter start={3} />
 >     <div class="h-5"/>
>      <a href="https://fresh.deno.dev" target="_blank">
>        <img width="197" height="37" src="/fresh-badge.svg" alt="Made with Fresh" />
>      </a>
>      <div class="text-sm text-gray-400">&copy; {new Date().getFullYear()} xuwt.com</div>
>    </div>
>  </>);
>}
> ```

I see our fresh origin code use constant function render html 

fresh/src/server/render.ts

![WX20220918-204445](https://user-images.githubusercontent.com/3103754/190902717-f0ec74b8-4b56-4348-89a4-0089d68810ac.png)

**Comment by hbsion** [comment]
> 如果您不提供生成相关页面的代码，我认为任何人都无法提供帮助，这比发布由华为托管的模棱两可的场外演示更有帮助。此外，您似乎正在收集 IP 地址，这在解决 github 问题时似乎很可疑。我是新来的，但我想这不是寻求指导的最佳方式。
> 
> 该`<Head>`组件在我的开发设置中按预期工作。

i am sorry 
the website is test cloudflare fresh can get real ip
the website is my test demo web

**Comment by irvanmalik48** [comment]
> I don't think anybody can help if you don't provide the code that generated the page in question, which would be much more helpful than posting an ambiguous off-site demo hosted by Huawei. Also, it looks like you're collecting IP addresses, which seems like a fishy thing to do while troubleshooting a github issue. I'm new here, but I imagine this isn't the best way to seek guidance.
> 
> The `<Head>` component works as expected in my dev setup.

It appears that setting the `charset` and `viewport` meta does create duplicate meta tag. I don't know if this is intentional whatsoever but any further explanations are welcome.

**Comment by connorlogin** [comment]
Good catch, I didn't see that before. It looks like Fresh renders those for you by default, probably because they're so standard across the web. It does that in the `server/render.ts` file: https://deno.land/x/fresh@1.1.1/src/server/render.ts?source#L338.

I don't know the implications of having duplicate viewports or charset tags (which one takes effect?), but I agree that's a little weird and feels wrong. I think the correct behavior should be to detect when those tags are manually specified in a `<Head>` component and not render the defaults when that happens.

I'm looking for ways to contribute and I can't find any other issues about this, so I'll fork Fresh and play around with fixing this quirk. I'll report back here if I make any progress.

**Comment by cnnrsprs** [comment]
Good catch, I didn't see that before.

I can't find any other issues about this and I'm looking for ways to contribute, so I'll fork fresh and play around with fixes. I'll report back on any progress made.

Update - Found something interesting, looks like Next also struggled with this at one point: https://github.com/vercel/next.js/issues/9794

**Comment by cnnrsprs** [comment]
Added a fix in #781. In that PR, the `template()` replaces the default charset/viewport with custom versions if they're specified. Waiting for review.

@hbsion Thank you for the code and screenshot you sent, they showed me where to start out.

**Comment by hbsion** [comment]
> Added a fix in #781. In that PR, the `template()` replaces the default charset/viewport with custom versions if they're specified. Waiting for review.
> 
> @hbsion Thank you for the code and screenshot you sent, they showed me where to start out.

Thank you，I wish fresh is very hot
